protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,424 @@
+            }
+            FromSwarm::ConnectionEstablished(ConnectionEstablished {
+                peer_id,
+                connection_id,
+                ..
+            }) => {
+                self.peers_to_handlers
+                    .entry(peer_id)
+                    .or_insert(connection_id);
+            }
+            FromSwarm::ConnectionClosed(ConnectionClosed {
+                peer_id,
+                connection_id,
+                ..
+            }) => {
+                self.handle_no_connection(peer_id, connection_id);
+            }
+            FromSwarm::DialFailure(DialFailure {
+                peer_id: Some(peer_id),
+                error,
+                connection_id,
+            }) => {
+                tracing::trace!("dialing {peer_id:?} failed: {error:?}");
+                self.handle_no_connection(peer_id, connection_id);
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: <Self::ConnectionHandler as ConnectionHandler>::ToBehaviour,
+    ) {
+        if matches!(event, Either::Left(_)) {
+            self.peers_to_handlers
+                .entry(peer_id)
+                .or_insert(connection_id);
+        }
+        match event {
+            Either::Right(nonce) => {
+                if self.pending_nonces.remove(&nonce) {
+                    tracing::trace!("Received pending nonce from {peer_id:?}");
+                } else {
+                    tracing::warn!("Received unexpected nonce from {peer_id:?}, this means that another node tried to be reachable on an address this node is reachable on.");
+                }
+            }
+            Either::Left(dial_request::ToBehaviour::PeerHasServerSupport) => {
+                self.inject_kown_server(peer_id);
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(Ok(TestEnd {
+                dial_request: DialRequest { nonce, .. },
+                reachable_addr,
+            }))) => {
+                if self.pending_nonces.remove(&nonce) {
+                    tracing::debug!(
+                        "server reported reachbility, but didn't actually reached this node."
+                    );
+                    return;
+                }
+                self.pending_events
+                    .push_back(ToSwarm::ExternalAddrConfirmed(reachable_addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(Err(err))) => {
+                match err.internal.as_ref() {
+                    dial_request::InternalError::FailureDuringDialBack { addr: Some(addr) }
+                    | dial_request::InternalError::UnableToConnectOnSelectedAddress {
+                        addr: Some(addr),
+                    } => {
+                        self.pending_events
+                            .push_back(ToSwarm::ExternalAddrExpired(addr.clone()));
```

Comment:
I believe there is a misunderstanding here. We should only emit `ExternalAddrExpired` for addresses that we have previously confirmed using `ExternalAddrConfirmed`. These events should always come as a pair. The condition under which we emit `ExternalAddrExpired` is a **re-testing** of an already confirmed external address that comes back as failed. To simplify things, we can omit the re-validation logic for now. Landing a version of AutonatV2 without re-validation is still very useful and it can easily be implemented later. Up to you :)

Instead of emitting `ToSwarm::ExternalAddrExpired` for an unsuccessful probe, we should emit a status update with the error.

Reply:
Ok I will change that.